cpp/src/copying/contiguous_split.cu

```diff
@@ -141,28 +279,164 @@ struct column_copy_functor {
+   cudf::column_device_view   offsets;
+};
+
+/**
+ * @brief Functor called by the `type_dispatcher` to collect string columns.
+ */
+struct string_column_collector {
+   template <typename T>
+   void operator()(cudf::column_view const& c, thrust::host_vector<column_preprocess_info>& offset_columns, size_type column_index){}   
+};
+template<>
+void string_column_collector::operator()<string_view>(cudf::column_view const& c, thrust::host_vector<column_preprocess_info>& offset_columns, size_type column_index)
+{
+   // constructing device view from the offsets column only, because doing so for the entire
+   // strings_column_view will result in memory allocation/cudaMemcpy() calls, which would
+   // defeat the whole purpose of this step.
+   cudf::column_device_view cdv((strings_column_view(c)).offsets(), 0, 0);
+   offset_columns.push_back(column_preprocess_info{column_index, c.offset(), c.size(), c.nullable(), cdv});      
+}
+
+/**
+ * @brief Preprocess information about all strings columns in a table view.
+ * 
+ * In order to minimize how often we touch the gpu, we need to preprocess various pieces of information
+ * about the string columns in a table as a batch process.  This function builds a list of the offset
+ * columns for all input string columns and computes this information with a single thrust call.  In addition,
+ * the vector returned is allocated for -all- columns in the table so further processing of non-string columns
+ * can happen afterwards.
+ * 
+ * The key things this function avoids
+ * - avoiding reaching into gpu memory on the cpu to retrieve offsets to compute string sizes.
+ * - creating column_device_views on the base string_column_view itself as that causes gpu memory allocation.
+ */
+thrust::host_vector<column_split_info> preprocess_string_column_info(cudf::table_view const& t, thrust::device_vector<column_split_info>& device_split_info, cudaStream_t stream)
+{          
+   // build a list of all the offset columns and their indices for all input string columns and put them on the gpu   
+   thrust::host_vector<column_preprocess_info> offset_columns;
+   offset_columns.reserve(t.num_columns());  // worst case
+   size_type column_index = 0;
+
+   // collect only string columns
+   std::for_each(t.begin(), t.end(), [&offset_columns, &column_index](cudf::column_view const& c){
+      cudf::experimental::type_dispatcher(c.type(), string_column_collector{}, c, offset_columns, column_index);
+      column_index++;
+   });
+   thrust::device_vector<column_preprocess_info> device_offset_columns = offset_columns;
+
+   // compute column split info for all string columns   
+   auto *sizes_p = device_split_info.data().get();
+   thrust::for_each(rmm::exec_policy(stream)->on(stream), device_offset_columns.begin(), device_offset_columns.end(),
+      [sizes_p] __device__ (column_preprocess_info cpi){
+         auto num_chars = cpi.offsets.head<int32_t>()[cpi.offset + cpi.size] - cpi.offsets.head<int32_t>()[cpi.offset];
```

Comment:
I'll give it a try but I can't see how that will be faster.  The standard test case here is 512 columns.  That'll be 1000 cudaMemcpys vs 1 kernel call and 1 cudaMemcpy.

Reply:
Update.  It's considerably slower.   Updated code (all host-side, memcpying the offsets)

```
 thrust::host_vector<column_split_info> split_info(t.num_columns());    
   std::for_each(offset_columns.begin(), offset_columns.end(),
      [&split_info] (column_preprocess_info const& cpi){
         int32_t offset_end, offset_start;
         cudaMemcpy(&offset_end, &cpi.offsets.head<int32_t>()[cpi.offset + cpi.size], sizeof(int32_t), cudaMemcpyDeviceToHost);
         cudaMemcpy(&offset_start, &cpi.offsets.head<int32_t>()[cpi.offset], sizeof(int32_t), cudaMemcpyDeviceToHost);

         auto num_chars = offset_end - offset_start;
         split_info[cpi.index].data_buf_size = cudf::util::round_up_safe(static_cast<size_t>(num_chars), split_align);         
         split_info[cpi.index].validity_buf_size = cpi.nullable ? cudf::bitmask_allocation_size_bytes(cpi.size, split_align) : 0;         
         split_info[cpi.index].offsets_buf_size = cudf::util::round_up_safe((cpi.size+1) * sizeof(size_type), split_align);
         split_info[cpi.index].num_chars = num_chars;
         split_info[cpi.index].chars_offset = offset_start; // cpi.offsets.head<int32_t>()[cpi.offset];
      }
   );
   return split_info;
```
(from gbenchmarks)
Original method, 512 columns, 4GB total data, no validity :    **444 ms**
New method, 512 columns, 4 GB total data, no validity :        **2592 ms**

Reply:
I was suggesting doing a bulk copy of all of the offset values at once. Effectively what the current kernel is doing, but moving _only_ the offset values, not the entire `column_split_info`.

Then move the computation of the contents of `column_split_info` to the host.

So something like:
```c++
thrust::device_vector<thrust::pair<size_type, size_type>> offsets(t.num_columns());
thrust::transform(offset_columns.begin(), offset_columns.end(), offsets.begin(), [](auto const& cpi){
return thrust::make_pair(cpi.offsets.head<int32_t>()[cpi.offset],cpi.offsets.head<int32_t>()[cpi.offset + cpi.size]);
}
thrust::host_vector<thrust::pair<size_type, size_type>> host_offsets(device_offsets);
// now compute column_split_infos using host_offsets
```

Reply:
I see.  Total amount of data moved between cpu/gpu is smaller if it's just the pair.  I'll give it a shot.

Reply:
Right, so times are effectively identical.   I assume the preference would be to use this new method so that I can eliminate the new rounding and bitmask size functions?

Reply:
Yeah, I expected performance to be pretty much the same, just a little be cleaner.